### DISCUSSION TITLE
Difficulty in mocking an async function from external library

### DISCUSSION THREAD

**Discussion by vaibhav6496** [cypress-io/cypress#28090] [Questions and Help] [answered]
I've a function used in my Angular app from a library: '@sec/blindfold'. The function name is `encryptStr`. It's an async function with signature as below:
`async function encryptStr(data: EncryptInput): Promise<StringResult>` where `StringResult` is an interface below:

```
export interface StringResult {
	result: string | null;
        error: Error | null;
}
```

The `encryptStr` calls bunch of external APIs, so instead of intercepting those calls I tried to stub the `encryptStr` in my cypress tests like below:

```
import * as blindFoldUtils from '@sec/blindfold';
describe('Certificates functionality', () => {
	let initCertValue, initKeyValue;
	const encryptKeyValue = 'somesecretValuefortesting';
	beforeEach(() => {
            cy.stub(blindFoldUtils, 'encryptStr').resolves({ result: encryptKeyValue, error: null });
     });
});
```

However, I can still see the original `encryptStr` is being called as the network tab shows call to the external APIs. Am I missing something? How to stub such an external async function's response in cypress?

**Comment by portyu9** [comment, marked as answer]
Hi @vaibhav6496 

The issue is not that `encryptStr()` is async. The important part is **which function reference your application is actually calling**.

This:

```js
import * as blindFoldUtils from '@sec/blindfold'

cy.stub(blindFoldUtils, 'encryptStr')
```

replaces `encryptStr` on the module object imported by the Cypress spec. It does not necessarily replace the static import/reference that was already bundled into your Angular application.

That is why the real `encryptStr()` can still run and make its external API calls.

For an **E2E test**, the cleanest boundary is usually to intercept the network requests made by `encryptStr()` with `cy.intercept()`.

For a **component/unit-style test**, wrap the external library behind an Angular service that your application injects, for example:

```ts
class EncryptionService {
  encrypt(data) {
    return encryptStr(data)
  }
}
```

Then provide a stub implementation when mounting the component/service under test. Cypress Angular component tests support overriding module-level providers.

That gives you an explicit dependency boundary rather than trying to mutate an imported ESM binding after the application has already imported it.

So `resolves()` is fine here; the problem is that the application is not calling the particular function reference you replaced.

Please mark this answer as accepted if it helped, thank you.
